### PR TITLE
Add root finding ode subclass. (not using the native solver's capabilities, just retracing and bisecting **in python**)

### PR DESCRIPTION
**Pull request by Krastanov**
No tests and not much documentation right now.

This is supposed to be a super simple root finding subclass of `ode`. If there is interest in this code I would be happy to provide tests and documentation.

**not using the native solver's capabilities, just retracing and bisecting in python**

### CODE REVIEW

**Comment by aarchiba** [comment]
Some of the ODE integrators natively support this sort of stopping conditions. But more generally, I don't think this is a good interface that allows efficient evaluation. This operation - integrate until a function hits a value - can be done with reasonable efficiency only for some integrators. These are the ones that support "dense output": normally, differential equation integrators only permit integration in one direction, call it "forward"; the integrator takes discrete steps forward, and these can be shortened but not safely lengthened. "Dense output" means that you can efficiently evaluate the solution anywhere within the last step. So the way to do this is, repeatedly let the integrator take whatever steps it wants, evaluating the condition at the end of the step. When it passes the desired value (conventionally zero), that last step provides a bracketing interval suitable for brent, and the integrator's dense output permits easy evaluation within that last step.

I have some code to do this, but for complicated reasons it's cython calling into a C++ DE solver. Nevertheless I think we should think more about the interface to use for this kind of technique.

**Comment by aarchiba** [comment]
Looking at the source code in scipy.integrate, I think explicit handling of dense output would need to be added first (which is a good idea regardless). But then it would be straightforward to add code in integrate() which would provide a stopping criterion for any integrator that supports dense output. 

Wouldn't hurt to have some new solvers in there too - the only reentrant ones don't seem to work for stiff problems.

**Comment by Krastanov** [comment]
1. This is indeed a very naive implementation that does NOT use the efficient root finding capabilities of the native-code solvers. Doing that would require actual new wrapping code and some more work.

2. Instead, the current code is the simplest possible solution, and also the one being suggested for this type of problem if you look at the mailing list over the last few years. (see http://mail.scipy.org/pipermail/scipy-user/2010-March/024890.html). I do agree though that backtracking and reseting is very inefficient.

3. The "API" here is the simplest I could think of, but I am open to criticism.

In conclusion, if the consensus is that the naive "no-new-wrapping-code" solution presented here is a bad idea, I would be glad to close this pull request. (However, I would like to point out that this feature has been asked for for a few years and new wrapping code has not be written yet and I do not know enough of the scipy innards to volunteer to do it.)

**Comment by aarchiba** [comment]
The main thing to be careful of is the user-visible interface, because once that gets released, we're stuck with it. I think the simplest interface I can picture is to simply modify the integrate method to take an additional argument, a function, and to stop when that function changes sign if that happens before the stopping time. (*Multiple* stopping functions is a possible wishlist item - for example, stop when the trajectory leaves a box - but could use the same interface, with a function returning a vector instead of a scalar.)

If for the moment that has to be implemented by inefficient resetting, so be it. But I'm not sure that the additional code required is a substantial challenge, particularly if you accept that not every integrator has dense output and can support this.

**Comment by Krastanov** [comment]
Thanks for the feedback. I will do the suggested modification (but I will not try my hands at the efficient solution) and let you guys decide whether it is worth pursuing after that. (I will probably do it later in the week/weekend).

(I only now noticed that you are the same person that wrote the old email I quoted. Sorry for the redundancy, but thanks anyway.)

**Comment by aarchiba** [comment]
As it currently stands, only vode and zvode support the step method, but both of them can backtrack within the last step without any additional wrapping of compiled code: you simply call the run method with a step backwards. lsoda is also capable of this without additional wrapping code, with essentially the same interface, but for some reason the single-step option is not provided from python; it'd be exactly the same code as for vode/zvode. The Dormand-Prince integrators have a separate not-yet-wrapped function for dense output; someone would have to write an f2py spec. But even just a start using vode would give people a way to do this cleanly, and establish an interface for doing this cleanly.

I put some better single-step and dense output handling code in https://github.com/aarchiba/scipy/tree/ode_stop but I'm still not sure of the cleanest interface. Really I'd prefer to use single integrator steps throughout, using dense output to give the user their values, but integrating this cleanly with the existing interface is awkward.

**Comment by WarrenWeckesser** [comment]
FWIW: For the `lsoda` wrapper, it might be nice to replace `lsoda` with `lsodar`, a version of `lsoda` with root-finding implemented in the Fortran code.  `lsoda` is also the backend of `odeint`, so it is also conceivable that `odeint` could be changed to use `lsodar`, with a `rootfunc` (or whatever) argument added. In fact, this has been done before, but not as part of scipy.  (If I find links to that code, I'll add it here.)  Unfortunately, I don't have time at the moment to work on this myself.

**Comment by aarchiba** [comment]
I believe I wrote a lsodar wrapper,  back in the mists of time, and there's one in pydstool. But I think a better solution is a clean interface to solvers with dense output (which is all of the ones we have, though some require a little extra wrapping), so that a generic scipy root-finder can be used. Since conditions and right-hand sides are forced to go through python, the extra cost of doing the root-finding in python is usually modest.

**Comment by WarrenWeckesser** [comment]
@aarchiba That's sounds fine to me.  The most important first step is deciding on a reasonable API.

Speaking of which, are y'all familiar with Matlab's event location API (http://www.mathworks.com/help/matlab/ref/odeset.html#zmw57dd0e478447)?  I won't suggest copying the API, but I will put in a request for similar functionality.  In Matlab, one can choose to detect only events where the function passes through the root in a specific direction (i.e. increasing or decreasing).  One can also tell the solver to *stop* when a certain event is detected.  (I guess this option would be available automatically when using scipy's `ode` class, which *requires* the user to write their own loop [*grumble, grumble*].  So an *isterminal* flag would probably make more sense for a future enhancement of `odeint`.)

**Comment by aarchiba** [comment]
Well, the "reasonable API" is a whole kettle of fish. I actually find the current arrangement quite a mess, but to some extent we are locked into it by compatibility. But I implemented an integrator object for another project (with a bunch of strange and unusual constraints, so it's not very general code). The general structure is:

Inner integrator object: 
* Setup specifies RHS, initial conditions, direction, error tolerance
* Step function advances by one integrator-chosen step
* Attributes reveal t,y at beginning and end of step
* Dense evaluator allows y to be computed for any t within the last step

Outer driver object:
* Setup specifies RHS, initial conditions, direction, error tolerance, event functions
* advance function moves the solution forward until the next time point or event
* Attributes reveal time and system state
* run function produces an array of system states

In my case, the run function takes a list of values for one of the event functions and produces a value every time the (increasing) event function crosses an input value. In general, you'd want to be able to specify a variety of different stopping criteria (event function reaches a value, time reaches a value) and the return array would specify why each data point was included (optionally, some events could cause the solver to actually stop rather than just returning a point). But the point is that the outer driver object provides event detection and odeint()-like functionality, while the inner object provides a relatively thin wrapper around the underlying solver. 

It might be possible to provide such a setup without much changing the current user-visible interface. scipy.integrate.ode could become the outer object, while the current non-public solver-wrapping objects could be restructured to be like the inner objects. It would mean moving the integrator loop to python, but since the RHS evaluations always go through python anyway, I don't think that's necessarily a problem.

**Comment by pv** [comment]
Re: API. My view here is that it would be easiest to do the redesign from scratch, in order not to be forced too much by suboptimal design decisions (different argument order for ode vs odeint, outer loop must be written in Python, odeint & ode are fully separate, output array not provided as argument, no way to write RHS in compiled language). The `odeint` API is simple enough that it could be retained, but `ode` would seem best to be replaced.

I think we should avoid assuming RHS evaluations go through Python. While this may be the most usual case, we should provide some way to write the callable in Cython etc. For example, integrate.quad can take ctypes callables and skip overheads (which we use in Scipy's elliptic harmonic special functions for significant speedup). This is not super-convenient, but I think there's plenty of room for improvement in this direction, and trying to find a reusable solution here would likely benefit the whole "ecosystem". For example, Cython already seems to provide enough information on its exported function objects that they could be dealt with.

**Comment by aarchiba** [comment]
Ah. Perhaps this discussion needs to go to the mailing list? It seems like the wrapping effort to provide one (let alone many) solvers with the ability to use compiled RHS functions would be substantial, but if it's feasible then one needs to worry more about the speed of the driver code. But indeed if ODE evaluation can be made fast that opens a variety of possibilities. I'm still inclined to use the underlying solvers in a one-step-at-a-time mode with dense evaluation, and just move the driver code to cython when it needs to be fast.

**Comment by bmcage** [comment]
In my view, vode, zvode and lsoda must be deprecated and replaced by their successor, cvode (yes, no true z form then, you need to split it up).
Wrappers out there have possibility to pass cython rhs, eg https://github.com/bmcage/odes/blob/master/scikits/odes/sundials/cvode.pyx

Also their, rootfn is used to pass root function capability. You really want to use that and not some higher level optimization loop which you cannot trust.

**Comment by aarchiba** [comment]
Note that there is a thread on scipy-dev discussing interfaces to ODE solvers, which might be a better place for dicsussion.

First: Why do you think the current solvers need to be deprecated? Do you have examples of problems where they fail? It appears that CVODE uses the same underlying approach (Adams-Moulton/BDF). It also appears that ZVODE takes advantage of analyticity, at least for stiff problems; is splitting into real problems as effective?

Second: Why do you think these solvers' built-in root-finding algorithms are better than those in scipy? Implementing these in scipy using the (existing or new) solvers' dense output allows a uniform flexible interface to stopping conditions. For example: there can be many conditions which trigger the solver to stop, they can either cause the solver to actually stop or simply log values, they can be restricted to sign changes in a particular direction, or they can occur whenever a function takes one of a long list of possible values. Using FORTRAN-level solvers' stopping conditions may well offer only some of these possiblities, and the subset will depend on the solver. On the other hand, scipy already has good root-finding routines, implementing specialized ones is not difficult, and dense output is a valuable tool for other reasons.

**Comment by bmcage** [comment]
CVODE (sundials) )is maintained. Long ago I used several, CVODE outperforms them in accuracy. So it's a matter of a more modern, more tested, and improved codebase. CVODE was developed by the same people who originally wrote LSODE/A and VODE based on the lessons learned. No ZVODE might be problematic to replace, I have no experience with that. I would suggest writing Alan C. Hindmarsh to ask why the functionality is not present in current sundials and what he would use for those type of problems today.

As to the root-finding, I know from discussions on the sundials mailing list that rootfinding in ode solvers is a difficult thing to get correct. Algorithm used in https://computation.llnl.gov/casc/sundials/documentation/cv_guide.pdf section 2.4, so also a secant method.
The important part however is that this is checked every internal step of the solver, not the output steps. I interpret this pull request as checking event on the output steps, which need not show sign changes in the events.

**Comment by aarchiba** [comment]
I recommend the thread on scipy-dev: http://thread.gmane.org/gmane.comp.python.scientific.devel/19752

As described in that PDF, the root-finding is not actually very complicated; it acts just as I was describing, using dense output within each (internal) solver step to locate roots with a modified secant algorithm. (Actually if derivatives can be obtained from dense output a careful Newton-type root finder might be faster.) What to do with those roots is a more complicated issue; I would rather use Cython to provide a uniform interface for root-finding for several different solvers, in case users want to choose the solver that works best for their problem. This also dramatically simplifies the code required to interface with a new compiled solver.

I have had a couple of tricky ODE-solution problems, including one I am currently working on, and I would like to ensure that scipy's ODE tools provide enough generality to solve at least them (modulo some headaches to do with high-precision internal variables, which I don't think we want to get into in scipy).

**Comment by bmcage** [comment]
As always, email threads go in tangent directions. I'm commenting here because of that mailing thread.
This pull request here does not allow me to understand the root finding is on the internal steps. Perhaps it should be expanded with test and example
THere are two things important:
1. speed: you don't want to force using single step in python to do root finding. So working with normal output and having root finding is best. If this mechanism links to the internal steps somehow, then fine
2. multiple root functions: you don't want to determine independently several roots based on the code in this pull request. It should happen in global fashion as in the manual of cvode pointed to earlier, eg the rhs function might change after a first root. 

As to solvers that are good for a problem. For stiff problems, not a single one in scipy is good. Allowing users to choose at the moment is allowing them to choose between different bad solutions. I started odes scikit development within a project because of that. The solvers present are a disservice to the users compared to eg Matlab. As indicated by people in the email thread, I'm not sure there is a good one for non-stiff problems either. 
Therefore, I would advocate moving forward to one good stiff solver, and one good adaptive timestepping RK solver. The good stiff solvers have root finding, no need to add something new then.
For more complex uses, there are specific ODE python packages that interface special case solvers. I don't think scipy should worry about those.

**Comment by Krastanov** [comment]
To clear up some confusion about the PR I updated the title and description. I will try to finish this PR this week.

**Comment by aarchiba** [comment]
The reason for the thread and not the PR is that this PR is taking the wrong approach. So discussing its specifics is not productive. 

Do you have a reference that shows LSODA or ZVODE are not good for stiff problems? They both include automatic method switching and support the same underlying BDF approach CVODE does. I do think we need a B-S or similar solver - not available from CVODE, and so needing a root-finding implementation. 

Also discussed in the email thread is writing the driver in cython, and setting up ways to get the RHS and stopping functions to be called without passing through the python interpreter. So speed need not be a problem. 

So please, contribute to the email thread instead.
